### ISSUE TITLE
AudioVAEDecoder.decode() missing output shape adjustment (_adjust_output_shape)

### ISSUE THREAD

**Issue by ziyu4huang** [dgrauet/ltx-2-mlx#36]
## Description

While investigating audio quality in the two-stage pipeline, I noticed that `AudioVAEDecoder.decode()` in `audio_vae.py` is missing the output shape adjustment that the upstream reference implementation performs.

## Upstream Reference

The upstream `AudioDecoder.forward` (in [ltx-core/audio_vae.py](https://github.com/Lightricks/LTX-2/blob/main/packages/ltx-core/src/ltx_core/model/audio_vae/audio_vae.py)) calls `_denormalize_latents` which computes a `target_shape` based on the causal configuration:

```python
target_frames = latent_shape.frames * LATENT_DOWNSAMPLE_FACTOR
if self.causality_axis != CausalityAxis.NONE:
    target_frames = max(target_frames - (LATENT_DOWNSAMPLE_FACTOR - 1), 1)
```

Then after the decoder network runs, `_finalize_output` and `_adjust_output_shape` are called. `_adjust_output_shape` crops or pads the output to exactly match `target_shape`, handling the causal time offset.

## MLX Port

The `decode()` method (lines 292-329 of `audio_vae.py`) skips all of this. After `conv_out`, it immediately transposes and returns. There is no output shape adjustment.

## Impact

The decoded mel spectrogram may have the wrong number of time frames (off by `LATENT_DOWNSAMPLE_FACTOR - 1 = 3` frames) due to the causal padding offset. This propagates as a misalignment into the vocoder, which processes whatever time dimension it receives.

While this is unlikely to cause dramatically audible artifacts (the vocoder just processes slightly more/fewer frames), it means the MLX output shape does not match the upstream output shape exactly, which could cause subtle timing or quality differences.

## Suggested Fix

Port the upstream `_adjust_output_shape` logic into the MLX `decode()` method, computing the expected target frames and cropping/padding the decoder output accordingly.

**Comment by dgrauet** [comment]
Thanks for the careful read of the upstream decoder, @ziyu4huang — you're right that upstream has `_adjust_output_shape` and that the MLX `decode()` doesn't call an equivalent final step. But after tracing the arithmetic and verifying empirically, **the MLX output shape already matches the upstream final output exactly**, so there is no off-by-3 (or off-by-anything) and nothing to fix. Details below.

## 1. The causal offset is absorbed per-upsample, not at the end

Upstream's `Upsample.forward` does `interpolate(2x)` → causal conv → **drop the first element on the causal axis**:

```python
# ltx-core upsample.py
x = torch.nn.functional.interpolate(x, scale_factor=2.0, mode="nearest")
x = self.conv(x)
case CausalityAxis.HEIGHT:
    x = x[:, :, 1:, :]   # length goes n -> 2n-1, keeping the "1 + 2*n" form
```

The MLX `AudioUpsample.__call__` is a line-for-line port (time is the H axis in our NHWC layout):

```python
x = mx.repeat(x, 2, axis=1)   # H (time)
x = mx.repeat(x, 2, axis=2)   # W (freq)
x = self.conv(x)
x = x[:, 1:, :, :]            # drop first time row — same as upstream x[:,:,1:,:]
```

Two upsamples on the time axis: `T → 2T-1 → 4T-3`. That is **exactly** upstream's target:

```python
target_frames = frames * LATENT_DOWNSAMPLE_FACTOR          # 4T
if causality_axis != NONE:
    target_frames = max(target_frames - (LATENT_DOWNSAMPLE_FACTOR - 1), 1)  # 4T - 3
```

So upstream's *raw* decoder output is also `4T-3`, and `_adjust_output_shape` is a **no-op safety net** for it (crop to `min(4T-3, 4T-3)`, pad 0). It only ever does work for variable-length / non-causal edge cases that don't occur in the LTX-2.3 config.

## 2. The config is causal-on-time, which is the case we match

I checked the real `config.json` shipped with the LTX-2.3 weights (q8/q4/bf16):

```
audio_vae.model.params.ddconfig.causality_axis     = 'height'   ← time is the causal axis
audio_vae.model.params.ddconfig.mid_block_add_attention = False
audio_vae.model.params.ddconfig.attn_resolutions   = []
audio_vae.model.params.ddconfig.ch_mult            = [1, 2, 4]
audio_vae.model.params.ddconfig.mel_bins           = 64
audio_vae.model.params.ddconfig.out_ch             = 2
```

(If this were `width`, time would be non-causal → `4T` and we *would* diverge. It isn't.)

## 3. Empirical: MLX output == upstream target, and `_adjust_output_shape` is a true no-op

Decoded the MLX `AudioVAEDecoder` across a T sweep, then ran a faithful NumPy port of upstream's `_adjust_output_shape` on the **actual** MLX output:

| latent T | MLX mel time | 4T−3 | `_adjust_output_shape(out)` == out |
|---:|---:|---:|:--:|
| 1 | 1 | 1 | ✅ identical |
| 2 | 5 | 5 | ✅ identical |
| 3 | 9 | 9 | ✅ identical |
| 4 | 13 | 13 | ✅ identical |
| 5 | 17 | 17 | ✅ identical |
| 9 | 33 | 33 | ✅ identical |
| 16 | 61 | 61 | ✅ identical |
| 33 | 129 | 129 | ✅ identical |
| 50 | 197 | 197 | ✅ identical |

Freq is `16 → 32 → 64` (width is non-causal → no drop) = `mel_bins=64`, channels = `2`. All three target axes match, so the crop-then-pad in `_adjust_output_shape` changes nothing.

## 4. End-to-end generation (real q4 weights)

`generate --distilled` at 384×256×49, joint audio:

| | value |
|---|---|
| audio stream | 48 kHz stereo |
| level | mean **−15.2 dB**, max **−1.4 dB** (healthy; cf. the −52 dB silent-audio bug in #34 — unrelated, already fixed in v0.14.10) |
| audio vs video duration | **2.01 s** vs **2.00 s** (in sync) |

A 3-frame mel offset (~30 ms) would show up as a wrong-length / desynced track; it doesn't.

## Conclusion

The MLX `decode()` produces a bit-shape-identical result to upstream's final (post-`_adjust_output_shape`) output for the shipped causal config. Porting `_adjust_output_shape` would add a no-op defensive crop/pad, not fix a divergence. Closing as not-a-bug.

We may still add a cheap `assert out.shape[2] == 4*T - 3` guard in `decode()` as a future regression tripwire — tracked separately if so. Thanks again for the report; the upstream cross-check was useful.
